modules/standard/IO.chpl

```diff
@@ -6122,6 +6122,114 @@ proc channel._read_complex(width:uint(32), out t:complex, i:int)
 }
 
 
+// This is called from within a param for loop in writef. It is important to
+// note here that the main benefit of this helper is that we don't pass all the
+// arguments from writef. This way, we can use the same code for an `arg` type
+// for which we have already created and instantiation of this.
+proc channel._writefOne(fmtStr, ref arg, i: int,
+                        ref cur: size_t, ref j: int,
+                        ref r: unmanaged _channel_regexp_info?, ref argType,
+                        ref conv: qio_conv_t, ref gotConv: bool,
+                        ref style: iostyle, ref err: syserr, origLocale: locale,
+                        len: size_t) throws {
+  gotConv = false;
+
+  if j <= i {
+    _format_reader(fmtStr, cur, len, err,
+                   conv, gotConv, style, r,
+                   false);
+  }
+
+  _conv_helper(err, conv, gotConv, j, argType);
```

Comment:
Passing argType as a tuple here makes `_writefOne` be generic on the size of argType (number of arguments) rather than just on the main generic thing `arg`. We could change `argType`s declaration from `var argType:(k+5)*c_int;` to `c_array(c_int, k+5)` and then pass a `argType: c_ptr(c_int)` to this function. The `c_ptr(c_int)` should behave similarly enough to a tuple to not require significant modifications. My only concern with that is that we might go beyond the end of the `argType` if there is some kind of program error / really weird format string that the code doesn't handle right. But that is easy to fix - we could pass the number of elements in `argType` as an additional argument and add bounds checking.